[Li KangLi] iP

lkldev opened the pull request:
### Not Duke but Luke

> "This is not going to go the way you think." - Luke Skywalker

Luke frees your mind of having to remember things you need to do using the force. It is

* text-based
* easy to learn
* ~~Fast~~SUPER FAST to use

All you need to do is,

1. Install [Java 11](https://www.oracle.com/java/technologies/downloads/).
2. download **Luke** from [here](https://github.com/lkldev/ip/releases/download/A-Jar/luke.jar).
3. run **Luke** with the following command: `java -jar luke.jar`
4. add your tasks.
5. let it manage your tasks for you :+1:

And it is __FREE__!

Features:

- [x] Managing tasks 
- [x] Managing deadlines
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice java too. Here's the main method:
```java
public class Luke {
    public static void main(String[] args) {
        Luke luke = new Luke("data/luke.txt");
        luke.start();
    }
}
```

Commit:
Add Gradle support

lkldev committed:
Add Greetings, Echo and Exit.

lkldev committed:
Add "Add" and "List" command

Update bot's personality

lkldev committed:
Add the ability to mark tasks as done

lkldev committed:
Remove toggleIsDone and add markAsDone and unmarkAsDone

lkldev committed:
Refractor Command to handle commands with arguments

Create ActionType to represent different Type of Commands

Rename CommandType to CommandAction to better suit Enum

Create Deadline, Event, Todo classes

Rename Bot from Duke to Luke

lkldev committed:
Add Text UI Testing

Change bot's personality

lkldev committed:
Add error handling to handle unknown commands, non integer input
and illegal arguments

Change CommandAction Enum to take in arguments required for each Command

Add static map of all the commands accepted to their corresponding enums

lkldev committed:
Update testing for error handling

Format text in README.md

lkldev committed:
Add delete command into bot

Update bot to use ArrayList instead of array

Update test cases to support delete operation

Update list to display different message if list is empty

Add error handling for index out of range.

lkldev committed:
Add "SAVE" into enums and hashmap

Prepare code to handle save command

lkldev committed:
Indent code to follow code style

lkldev committed:
Add saving to file and loading from file capability

Remove previous assumption of save command

Create ADT TaskList to hold all the Task

lkldev committed:
Add DateTimeParser to parse string to date and format date to string

Update testcases to reflect the new output and error messages

lkldev committed:
Merge branch 'master' into branch-Level-7

# Conflicts:
#	src/main/java/CommandAction.java
#	src/main/java/Luke.java

lkldev committed:
Merge branch 'branch-Level-7'

lkldev committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/Luke.java

Update test case to include saving and loading of data

lkldev committed:
Refactor code to make it more OOP
- Create Parser class to parse commands
- Split all commands into different classes
- Extract ui code to Ui class
- Simplify start logic in Luke

Update test case to reflect new output messages

lkldev committed:
Update datetime format to support more format, change datetime display

lkldev committed:
Put files into packages

lkldev committed:
Change access modifier for method in Luke

lkldev committed:
Add unit testing for parser class

Update text-ui-test to reflect new output

Update packaging for Luke.java

lkldev committed:
Add Manifest for jar release

lkldev committed:
Remove unnecessary files and change file path

lkldev committed:
Update code to follow coding standard and style

lkldev committed:
Add files for checkstyle

lkldev committed:
Add Find command and refactor tasks to allow filtering ability

Group FindCommand and ListCommand as ReadCommand to share methods

lkldev committed:
Merge branch 'branch-Level-9'

lkldev committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/luke/commands/CommandAction.java
#	src/main/java/luke/parser/Parser.java
#	src/test/java/luke/parser/ParserTest.java

Ensure merged files adhere to coding standards

lkldev committed:
Add Javadocs and remove unused variables

lkldev committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/luke/Luke.java
#	src/main/java/luke/commands/DeleteCommand.java
#	src/main/java/luke/commands/ExitCommand.java
#	src/main/java/luke/commands/ListCommand.java
#	src/main/java/luke/commands/MarkCommand.java
#	src/main/java/luke/commands/UnmarkCommand.java
#	src/main/java/luke/data/tasks/Deadline.java
#	src/main/java/luke/data/tasks/Event.java
#	src/main/java/luke/parser/Parser.java

lkldev committed:
Add missing documents for merged files

Ensure all files adheres to code style

lkldev committed:
Merge branch 'add-gradle-support' of https://github.com/lkldev/ip into branch-A-Gradle

lkldev committed:
Add gradle into project

lkldev committed:
Merge branch 'branch-A-Gradle'

blaketingyu commented on src/main/java/luke/commands/ActionType.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Types of actions that are performed by the commands.
+ */
+public enum ActionType {
```
Like how you've made use of ENUM for your different action types here! 👍

blaketingyu commented on src/main/java/luke/commands/MarkCommand.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class MarkCommand extends UpdateCommand {
+
+    private static final String DEFAULT_MESSAGE = "Using the force... Great! I have forced this task as done.";
```
I like how you name your constant values in CAPs. 👍

blaketingyu commented on src/main/java/luke/data/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+    public TaskList() {
+        taskList = new ArrayList<>();
+    }
+
```
Maybe you could add a method header for getData() method. 👍

lkldev replied:
Good suggestion! I will add that in in the next commit :)

blaketingyu commented on src/main/java/luke/data/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+        return taskList.remove(index);
+    }
+
+    public void setFiltered(Predicate<Task> predicate) {
```
Like how you've utilized the Predicate class to implement filter! 👍

blaketingyu commented on src/main/java/luke/parser/Parser.java:
```diff
@@ -0,0 +1,204 @@
+import java.util.HashMap;
+import java.util.Map;
+
+import luke.commands.AddCommand;
```
I like how you've listed the imports lexicographically! 👍

blaketingyu commented on src/main/java/luke/parser/Parser.java:
```diff
@@ -0,0 +1,204 @@
+     * @return The arguments mapped to the keywords of the command.
+     * @throws IllegalArgumentException If the argument does not fit the syntax of the command.
+     */
+    private static Map<String, String> parseAddArguments(CommandAction cmdAction, String[] args)
```
Really like how you've made use of hashmaps to parse the arguments from the user input. It's a pretty creative way to parse the user's argument! 👍

blaketingyu commented on src/main/java/luke/commands/ReadCommand.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Takes in a task list and return the appropriate message based on the items in the task.
+     *
+     * @param taskList   The task list to loop through.
```
I like how most of your variable names and method names are self-explanatory. The naming is sufficient for reader to know what is going on in your code.

blaketingyu reviewed (commented):
Overall, I found your code very neat and easy to read! I did not find much violations of Java coding standards and the Javadocs are really well written and succint. Perhaps, you could add in method headers for some of your methods. Other than that, it's really good overall! 👍

lkldev committed:
Add missing javadoc for TaskList method.

lkldev committed:
Add GUI for luke

Remove text UI

lkldev committed:
Merge branch 'branch-Level-10'

lkldev committed:
Update gradle to run launcher instead of luke

lkldev committed:
Add assertion to code

We assumed the state of our variables especially when using private
functions.

Let's add assertion to our private function so that if a wrong argument
is passed into our private function, it will be flag out by assertion.

Using assertion instead of exceptions as private functions are not
invoked by others other than methods in the own class. These scenarios
are considered logically impossible, hence if the assertion is raise,
it is likely due to a mistake made by ourselves.

lkldev committed:
Improve code quality

Code contains magic literals that is distracting the reviewer from the
actual implementation of the method.

Some codes also have the happy path nested, which is distracting as
those method are not the usual case.

Switch case default are being used to execute the last option, which
may cause unintended bugs.

Let's
* Ensure that there are no magic literals, replace all with constants
* Use guard clauses to ensure that happy path is prominent
* Remove unused methods and constants
* Use default branch, and to throw exception if there are no default
action

This would ensure that the code remains readable and easier to spot any
mistakes or bugs.

lkldev committed:
Merge pull request #1 from lkldev/branch-A-Assertions

Add assertion to code

lkldev committed:
Merge branch 'master' into branch-A-CodeQuality

lkldev committed:
Merge pull request #2 from lkldev/branch-A-CodeQuality

Improve code quality

lkldev committed:
Add yml file for github workflow

cashewnade commented on src/main/java/luke/commands/CommandResult.java:
```diff
@@ -0,0 +1,38 @@
+        return this.result;
+    }
+
+    public boolean isExit() {
```
Instead of just returning isExit, how about implementing it such that it returns whether the command is an instance of ExitCommand?

lkldev replied:
Hmm, this would create a cyclic dependency between Command and ExitCommand. Also, this class would only contain the results of the command, not the command type itself... It is used for the GUI to print out the result without knowing what command has been executed :) Thanks for the suggestion!

cashewnade commented on src/main/java/luke/commands/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    /**
+     * Takes in a task list, filter the task list based on keywords and read the task list.
```
The JavaDocs is good enough. Perhaps you can try to change it to "filters" and "reads". :)

lkldev replied:
Thanks for the suggestions!

cashewnade commented on src/main/java/luke/commands/ReadCommand.java:
```diff
@@ -0,0 +1,48 @@
+public abstract class ReadCommand extends Command {
+
+    /**
+     * Takes in a task list and return the appropriate message based on the items in the task.
```
The same comment, "returns" would be more appropriate! :)

cashewnade commented on src/main/java/luke/commands/ReadCommand.java:
```diff
@@ -0,0 +1,48 @@
+        String msg = emptyMsg;
+        boolean isEmpty = true;
+        //Since task list may be fully filtered but not empty, require this check to see if tasks are filtered
+        if (!taskList.isEmpty()) {
```
Perhaps you can leave a line here for better readability! :)

cashewnade commented on src/main/java/luke/parser/Parser.java:
```diff
@@ -0,0 +1,223 @@
+            throws IllegalArgumentException {
+        assert(cmdAction.getCommandActionType() == ActionType.ADD);
+        Map<String, String> argsMap = new HashMap<>();
+        if (args.length < 2) {
```
Perhaps add a line break between each if clause, to enhance readability! :)

cashewnade reviewed (commented):
LGTM! Your code follows most of the coding standards! A few grammatical issues in the documentation but otherwise everything looks good! Interesting use of HashMaps, and the CommandResults class! Look forward to seeing the full product!

lkldev committed:
Modify javadocs based on PR comments

Refactor Parser to handle argument more elegantly

Add extra test cases to ensure parser works as intended

lkldev committed:
Update test method name

lkldev committed:
Add Recurring task

Users will now be able to add recurring tasks using the following
format:
 * Syntax: `recur <task> /every day|week|month|year`
 * Examples:
    * `recur todo code /every day`
    * `recur event team meeting /at 13/02/2022 10:30 /every week`
    * `recur deadline assignment /by 18/02/2022 23:59 /every month`

Numerous test cases have been added to ensure that recur would function
as intended.

Refactor code for storage initialization to ensure compatibility with
recurring tasks.

Update Parser to handle both Add commands and Recur commands correct ly.

Update text displayed on GUI to improve readability.

lkldev committed:
Merge pull request #4 from lkldev/branch-BCD-Extension

Add Recurring task

lkldev committed:
Improve GUI of Luke

lkldev committed:
Improve GUI of Luke

lkldev committed:
Rename UI.png to Ui.png

lkldev committed:
Move Ui.png to docs

lkldev committed:
Update User Guide

lkldev committed:
Add line breaks in User guide for readability
